File: src/predef_java/BenefitService.java
package predef_java;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Set;

public interface BenefitService<AssignBenefitResult, UnassignBenefitResult, ScheduleResult> {
	List<UserResponse> getUsers(GetUsersFilter filter);

	AssignBenefitResult assignBenefit(Set<UserId> users, Benefit benefit);

	UnassignBenefitResult unassignBenefit(Set<UserId> users, Benefit benefit);

	ScheduleResult scheduleBenefitAssigning(Set<UserId> users, Benefit benefit, LocalDateTime at);
}

File: src/predef/BenefitService.scala
package predef

import java.time.LocalDateTime

enum GetUsersFilter {
  case HAS_BENEFITS, HAS_OR_WILL_HAVE_BENEFITS, HAD_BENEFITS
}

trait BenefitService[
    F[_],
    AssignBenefitResult,
    UnassignBenefitResult,
    ScheduleResult
] {
  def getUsers(filter: GetUsersFilter): F[Seq[UserResponse]]

  def assignBenefit(
      users: Set[UserId],
      benefit: Benefit
  ): F[AssignBenefitResult]

  def unassignBenefit(
      users: Set[UserId],
      benefit: Benefit
  ): F[UnassignBenefitResult]

  def scheduleBenefitAssigning(
      users: Set[UserId],
      benefit: Benefit,
      at: LocalDateTime
  ): F[ScheduleResult]
}
